native-assets/fractional-NFT/project/contracts/f-NFT-contract/src/main.sw

```diff
@@ -0,0 +1,475 @@
+    /// * When the asset is not shares to an NFT.
+    /// * When the amount sent isn't all shares of an NFT.
+    ///
+    /// # Examples
+    ///
+    /// ```sway
+    /// use src6::SRC6;
+    /// use std::constants::DEFAULT_SUB_ID;
+    ///
+    /// fn foo(vault_contract: ContractId, reciever: Identity, nft: AssetId, shares: AssetId) {
+    ///     let vault_abi = abi(SRC6, vault_contract);
+    ///     let _ = vault_abi {
+    ///         gas: 10000,
+    ///         coins: 100_000_000,
+    ///         asset_id: shares,
+    ///     }.withdraw(reciever, nft, DEFAULT_SUB_ID);
+    /// }
+    /// ```
+    #[payable]
+    #[storage(read, write)]
+    fn withdraw(
+        receiver: Identity,
+        underlying_asset: AssetId,
+        vault_sub_id: SubId,
+    ) -> u64 {
+        require(vault_sub_id == ZERO_B256, SubIdError::InvalidSubId);
+
+        let sent_amount = msg_amount();
+        let f_nft_asset_sub_id = sha256((underlying_asset, vault_sub_id));
+        let f_nft_asset = AssetId::new(contract_id(), f_nft_asset_sub_id);
+        require(msg_asset_id() == f_nft_asset, WithdrawError::InvalidAsset);
+        require(sent_amount == SHARES, WithdrawError::AllSharesNotReturned);
+
+        burn(f_nft_asset_sub_id, SHARES);
+        transfer(receiver, underlying_asset, 1);
+
+        log(Withdraw {
+            caller: msg_sender().unwrap(),
+            receiver,
+            underlying_asset,
+            vault_sub_id,
+            withdrawn_amount: SHARES,
+            burned_shares: SHARES,
+        });
+
+        1
+    }
+
+    /// Returns the amount of managed assets of the given asset.
+    ///
+    /// # Arguments
+    ///
+    /// * `underlying_asset`: [AssetId] - The NFT for which the amount of managed assets should be returned.
+    /// * `vault_sub_id`: [SubId] - The SubId of the vault.
+    ///
+    /// # Returns
+    ///
+    /// * [u64] - The amount of managed assets of the given asset.
+    ///
+    /// # Examples
+    ///
+    /// ```sway
+    /// use src6::SRC6;
+    /// use std::constants::DEFAULT_SUB_ID;
+    ///
+    /// fn foo(vault_contract: ContractId, nft: AssetId) {
+    ///     let vault_abi = abi(SRC6, vault_contract);
+    ///     let managed_assets = vault_abi.managed_assets(nft, DEFAULT_SUB_ID);
+    ///     assert(managed_assets == 1);
+    /// }
+    /// ```
+    #[storage(read)]
+    fn managed_assets(underlying_asset: AssetId, vault_sub_id: SubId) -> u64 {
+        if vault_sub_id != ZERO_B256 {
+            return 0;
+        }
+
+        match this_balance(underlying_asset) {
+            1 => 1,
+            _ => 0,
+        }
+    }
+
+    /// Returns the maximum amount of assets that can be deposited into the contract, for the given NFT.
+    ///
+    /// # Additional Information
+    ///
+    /// Maximum is `Some(1)` as NFTs only have a supply of one.
+    ///
+    /// # Arguments
+    ///
+    /// * `receiver`: [Identity] - The hypothetical receiver of the shares.
```

Comment:
Why hypothetical?

Reply:
This is part of the SRC-6 standard. It's a function to query the amount depositible by a user.